modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1925,29 +1922,41 @@ void GDScriptAnalyzer::resolve_return(GDScriptParser::ReturnNode *p_return) {
 		}
 		result = p_return->return_value->get_datatype();
 	} else {
-		// Return type is null by default.
 		result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
 		result.kind = GDScriptParser::DataType::BUILTIN;
 		result.builtin_type = Variant::NIL;
 		result.is_constant = true;
 	}
 
 	if (has_expected_type) {
-		expected_type.is_meta_type = false;
-		if (expected_type.is_hard_type()) {
-			if (!is_type_compatible(expected_type, result)) {
-				// Try other way. Okay but not safe.
+		if (expected_type.is_variant()) {
+			// variant expectation
+			if (!expected_type.is_hard_type()) {
+				// untyped expectation
+				mark_node_unsafe(p_return);
```

Comment:
This enters in our discussion "what's a green line". If a function does not declare any return type, it should be Variant by default, so returning anything is safe.

Reply:
Yes. To recap my opinion: the return is typed as Variant, yes, but weak one and it does not represent what user really wants. Unlike hard Variant which is a specified type with explicit intention, weak one is just a fallback in absence of a type but most likely does not reflect user's expectation. So since we don't know what type user has in mind it should be marked as unsafe. To make it safe they need to specify the type.

Reply:
Done. Removed that logic.